lib/awaitableSender.ts

```diff
@@ -0,0 +1,178 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache License. See License in the project root for license information.
+
+import {
+  Delivery, Message, Sender as RheaSender, SessionEvents
+} from "rhea";
+import * as log from "./log";
+import { BaseSender, BaseSenderOptions } from "./sender";
+import { SenderEvents } from "rhea";
+import { OnAmqpEvent, EventContext } from "./eventContext";
+import { Session } from "./session";
+import {
+  OperationTimeoutError, InsufficientCreditError, SendOperationFailedError
+} from "./errorDefinitions";
+
+/**
+ * @internal
+ */
+export interface PromiseLike {
+  resolve: (value?: any) => void;
+  reject: (reason?: any) => void;
+  timer: NodeJS.Timer;
+}
+
+/**
+ * Describes the event listeners that can be added to the AwaitableSender.
+ * @interface Sender
+ */
+export declare interface AwaitableSender {
+  on(event: SenderEvents, listener: OnAmqpEvent): this;
+}
+
+export interface AwaitableSenderOptions extends BaseSenderOptions {
+  /**
+   * The duration in which the promise to send the message should complete (resolve/reject).
+   * If it is not completed, then the Promise will be rejected after timeout occurs.
+   * Default: `20 seconds`.
+   */
+  sendTimeoutInSeconds?: number;
+}
+
+/**
+ * Describes the async version of the sender where one can await on the message being sent.
+ * @class AwaitableSender
+ */
+export class AwaitableSender extends BaseSender {
+  /**
+   * The duration in which the promise to send the message should complete (resolve/reject).
+   * If it is not completed, then the Promise will be rejected after timeout occurs.
+   * Default: `20 seconds`.
+   */
+  sendTimeoutInSeconds: number;
+  /**
+   * @property {Map<number, PromiseLike} deliveryDispositionMap Maintains a map of delivery of
+   * messages that are being sent. It acts as a store for correlating the dispositions received
+   * for sent messages.
+   */
+  deliveryDispositionMap: Map<number, PromiseLike> = new Map<number, PromiseLike>();
+
+  constructor(session: Session, sender: RheaSender, options: AwaitableSenderOptions = {}) {
+    super(session, sender, options);
+    this.sendTimeoutInSeconds = options.sendTimeoutInSeconds || 20;
+
+    const onSuccess = (delivery: Delivery) => {
```

Comment:
Can we rename `onSuccess` and `onFailure` to `onSendSuccess` and `onSendFailure`? Helps distinguishing `onFailure` and `onError`. The former is for each send operation, the latter is for the entire sender link, right?

Reply:
Yup. However, onError internally calls onFailure since the steps to be taken are by and large the same for `onFailure` and `onError`. Hope that doesn't confuse us with why onError is calling onSendFailure.